[GraphOptimizer] Sink Transpose below Arithmetic nodes when one operand is a Constant

SplitInfinity opened the pull request:
**Summary**
This commit extends the optimization that sinks `Transpose` nodes below `Add`
nodes to handle the case in which one of the operands of the `Add` node is
a `Constant`.

**Testing**
This commit adds a unit test that tests that a `Transpose` node is sunk below
an `Add` node when one operand is a `Constant` and the other is a `Transpose` node.

```
$ ./tests/GraphOptzTest --gtest_filter=*sinkTransposeBelowArithmeticNodesWithConstantOperand*
Note: Google Test filter = *sinkTransposeBelowArithmeticNodesWithConstantOperand*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from GraphOptz
[ RUN      ] GraphOptz.sinkTransposeBelowArithmeticNodesWithConstantOperand
[       OK ] GraphOptz.sinkTransposeBelowArithmeticNodesWithConstantOperand (5 ms)
[----------] 1 test from GraphOptz (5 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (5 ms total)
[  PASSED  ] 1 test.
```

SplitInfinity commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -445,6 +460,70 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
                                     SN->getValue());
           RTR = F->createTranspose("transpose", NS, LTR->getShuffle());
           changed = true;
+        } else if (isa<Constant>(node->getNthInput(ArithmeticNode::LHSIdx)) &&
```
I need to check that the `Constant` has only one use as well, right?

jfix71 replied:
Not checking isn't a correctness issue -- you're creating a new Constant and so other users should still have access to the un-transposed version. But there is a tradeoff here -- if it's a large Constant then we probably wouldn't want to make a duplicate. If it's a small Constant then maybe the compute tradeoff is worth the extra memory. For now (and if your use case only has one user) I'd say add the check.

SplitInfinity replied:
This is tricky thanks to the numerical correctness tests I added (hah!). The flow of a test is now like so:

1. Build graph.
2. Make pre-optimization assertions about the graph.
3. Clone the graph.
4. Optimize the cloned version of the graph.
5. Make post-optimization assertions about the optimized, cloned version of the graph.
6. Run both the original and optimized graphs with the same input and make sure the outputs are identical.

Because the graph is cloned and then its clone is optimized, these `Constants` actually have *two* uses: one in the original graph, and another in its clone.

jfix71 replied:
Hmm, yeah this is where it would be nice to have `Module::clone()` to use instead of just `Function::clone()`. Then we could copy the Constants in addition to the Function and they'd have the same number of users.

(Another place where this is a little annoying is unit tests for training, where we create two EEs/Modules: one for training and one for inference, and then we explicitly re-create the same Function twice on each. E.g. see MLTest `trainASimpleNetwork` and many others.)

SplitInfinity commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -445,6 +460,70 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
+          // Create Transpose on the LHS that has the same shuffle as RTR.
+          LTR = F->createTranspose("transpose", NC, RTR->getShuffle());
+          changed = true;
+        } else if (isa<Constant>(node->getNthInput(ArithmeticNode::RHSIdx)) &&
```
I need to check that the `Constant` has only one use as well, right?

SplitInfinity commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -916,6 +916,57 @@ TEST_F(GraphOptz, sinkTransposeBelowArithmeticNodes) {
 
+/// Check that Transpose node is sunk below arithmetic nodes when one of the
+/// operands is a Constant.
+TEST_F(GraphOptz, sinkTransposeBelowArithmeticNodesWithConstantOperand) {
```
Let's also test the case in which the `Constant` is on the RHS of the `Add`.

opti-mix commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -445,6 +460,70 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
+          // second Transpose. The statement below statically computes this
+          // first Transpose.
+          T.transpose(&NC->getPayloadMutable(),
+                      invertShuffle(RTR->getShuffle()));
```
This big snippet and a similar snippet below look almost identical. Move then into a helper function?

SplitInfinity commented:
Addressed comments:
- moved the code that adds a `Transpose` after the `Constant` operand of an arithmetic node into a helper function so that it can be reused by the cases of the `if-statement` that handle the `Constant` as the RHS/LHS operand
- extended test to test that the optimization works when the `Constant` is LHS and RHS of an `Add` node

nickgg commented:
@SplitInfinity did you push your update? I don't see that change.

SplitInfinity commented:
@nickgg 

Good call, I committed the changes locally but forgot to push.
